File: render/acts/a06_the_night.py
"""Act 06 · the night band. The only act that changes the surface."""

from copytext import t

from .. import html
from ..fmt import maybe


def strip(summary: dict) -> str:
    night_first = summary["night_first_week"]
    night_last = summary["night_last_week"]
    return html.kpis([
        {"label": t("night.kpi.first_week"),
         "value": f"{night_first:.0f} {t('unit.min')}"},
        {"label": t("night.kpi.last_week"),
         "value": f"{night_last:.0f} {t('unit.min')}",
         "delta": t("delta.times", n=summary["night_multiple"])},
        {"label": t("night.kpi.last_screen_first"),
         "value": summary["last_screen_first_week"]},
        {"label": t("night.kpi.last_screen_last"),
         "value": summary["last_screen_last_week"],
         "delta": shift(summary["last_screen_shift_min"])},
        {"label": t("night.kpi.first_unlock"),
         "value": summary["wake_last_week"],
         "delta": t("delta.minutes", n=summary["wake_shift_min"])},
        {"label": t("night.kpi.sleep_window"),
         "value": maybe(summary["sleep_last_week"], ".1f", t("unit.hours")),
         "delta": shift(sleep_change(summary))},
    ])


def shift(minutes) -> str | None:
    return None if minutes is None else t("delta.minutes", n=minutes)


def sleep_change(summary: dict):
    first, last = summary["sleep_first_week"], summary["sleep_last_week"]
    return None if first is None or last is None else (last - first) * 60


def reading(ctx) -> str:
    summary = ctx.profile["summary"]
    if ctx.user == "A":
        return html.note(t("night.note.user_a",
                           last_use=summary["last_use_mean"]), "good")
    return html.note(t(
        "night.note.drift",
        end_first=summary["last_screen_first_week"],
        end_last=summary["last_screen_last_week"],
        end_shift=summary["last_screen_shift_min"],
        wake_first=summary["wake_first_week"],
        wake_last=summary["wake_last_week"],
        wake_shift=summary["wake_shift_min"],
        sleep_first=summary["sleep_first_week"],
        sleep_last=summary["sleep_last_week"],
        sleep_loss=abs(summary["sleep_last_week"]
                       - summary["sleep_first_week"]) * 60,
        pick_first=summary["night_pickups_first_week"],
        pick_last=summary["night_pickups_last_week"]), "serious")


def build(ctx) -> str:
    return (
        strip(ctx.profile["summary"])
        + html.chart("night_drift", "shared")
        + reading(ctx)
        + html.grid(html.chart("day_span.night"),
                    html.chart("compare.night_pickups", "shared"))
        + html.note(t("night.weight.body", night=ctx.df.night_min.mean(),
                      screen=ctx.df.screen_min.mean()))
    )

File: web/acts/a06-the-night.ts
/**
 * Act 06 · the night band. The only act that changes the surface mid-page.
 */

import { explain } from "../copy/explain";
import { unit } from "../copy/units";
import { clock, maybe } from "../format";
import { chart, grid, kpis, note, type Kpi } from "../html";
import type { Summary } from "../types/index";
import type { Act, Context } from "./act";

const copy = {
  eyebrow: "23:00 to 06:00",
  title: "What happens at night",

  kpi: {
    firstWeek: "Late night, wk 1",
    lastWeek: "Late night, wk 4",
    lastScreenFirst: "Last screen, wk 1",
    lastScreenLast: "Last screen, wk 4",
    firstUnlock: "First unlock",
    sleepWindow: "Sleep window",
  },
  times: (n: number) => `×${n.toFixed(0)}`,
  minutes: (n: number) => `${n > 0 ? "+" : ""}${n.toFixed(0)} min`,

  driftNote: (endFirst: string, endLast: string, wakeFirst: string,
              wakeLast: string, pickLast: number, sleepLoss: number) =>
    `<b>You are losing sleep.</b> You started the month going to bed at `
    + `${endFirst} and ended it averaging ${endLast}, whilst always waking up `
    + `at the same time — ${wakeFirst} then ${wakeLast}. You went from hardly `
    + `unlocking your phone after midnight to doing it nearly `
    + `${pickLast.toFixed(0)} times a night. That is `
    + `<b>${sleepLoss.toFixed(0)} minutes less rest, every night</b>.`,

  userANote: (lastUse: string) =>
    `You only see user B's usage here, because A slept remarkably well throughout the whole of May: zero minutes between 23:00 and 06:00 on all thirty days, last screen at ${lastUse} on average.`,

  weightNote: (night: number, screen: number) =>
    "The night-time usage increase is the main reason your wellbeing score has dropped so much. "
    + `Sleep! It is the smallest number on this page — ${night.toFixed(0)} min a day against ${screen.toFixed(0)} of screen — and it carries 20 % of the index, because an hour at 01:00 comes out of your rest and an hour at 17:00 does not.`,

  next: "Now you know when. The next question is what.",
};

/** A shift in minutes, or nothing where there is no pair to compare. */
function shift(minutes: number | null): string | undefined {
  return minutes === null ? undefined : copy.minutes(minutes);
}

function sleepChange(s: Summary): number | null {
  const { sleep_first_week_h: first, sleep_last_week_h: last } = s;
  return first === null || last === null ? null : (last - first) * 60;
}

function hoursShift(first: number | null, last: number | null): number | null {
  return first === null || last === null ? null : (last - first) * 60;
}

function strip(s: Summary): string {
  const items: Array<[string, string, string | undefined]> = [
    [copy.kpi.firstWeek, `${s.night_first_week.toFixed(0)} ${unit.min}`,
     undefined],
    [copy.kpi.lastWeek, `${s.night_last_week.toFixed(0)} ${unit.min}`,
     copy.times(s.night_multiple)],
    [copy.kpi.lastScreenFirst, clock(s.last_screen_first_week_h), undefined],
    [copy.kpi.lastScreenLast, clock(s.last_screen_last_week_h),
     shift(hoursShift(s.last_screen_first_week_h, s.last_screen_last_week_h))],
    [copy.kpi.firstUnlock, clock(s.wake_last_week_h),
     shift(hoursShift(s.wake_first_week_h, s.wake_last_week_h))],
    [copy.kpi.sleepWindow, maybe(s.sleep_last_week_h, 1, unit.hours),
     shift(sleepChange(s))],
  ];
  return kpis(items.map(([label, value, delta]): Kpi =>
    (delta === undefined ? { label, value } : { label, value, delta })));
}

function reading(ctx: Context): string {
  const s = ctx.profile.summary;
  if (ctx.user === "A") {
    return note(copy.userANote(clock(s.last_use_mean_h)), "good");
  }
  const first = s.sleep_first_week_h ?? 0;
  const last = s.sleep_last_week_h ?? 0;
  return note(copy.driftNote(
    clock(s.last_screen_first_week_h), clock(s.last_screen_last_week_h),
    clock(s.wake_first_week_h), clock(s.wake_last_week_h),
    s.night_pickups_last_week, Math.abs(last - first) * 60), "serious");
}

export const act: Act = {
  id: "06",
  part: 2,
  eyebrow: copy.eyebrow,
  title: copy.title,
  next: copy.next,
  build(ctx: Context): string {
    const s = ctx.profile.summary;
    return strip(s)
      + chart("night_drift", explain("night_drift"), { scope: "shared" })
      + reading(ctx)
      + grid([chart("day_span.night", explain("day_span.night")),
              chart("compare.night_pickups", explain("compare.night_pickups"),
                    { scope: "shared" })])
      + note(copy.weightNote(s.night_mean, s.screen_mean));
  },
};
